detectable/src/main/java/com/synopsys/integration/detectable/detectables/bazel/pipeline/step/FinalStepJsonProtoHaskellCabalLibraries.java

```diff
@@ -0,0 +1,146 @@
+    }
+
+    private Dependency haskageCompNameVersionToDependency(final String compName, final String compVersion) {
+        final ExternalId externalId = (new ExternalIdFactory()).createNameVersionExternalId(hackageForge, compName, compVersion);
+        externalId.createBdioId(); // Validity check; throws IllegalStateException if invalid
+        return new Dependency(compName, compVersion, externalId);
+    }
+
+    private Optional<JsonObject> extractHaskellCabalLibraryRule(final JsonElement resultsArrayMemberElement) {
+        final JsonObject resultsArrayMemberObject = resultsArrayMemberElement.getAsJsonObject();
+        final JsonElement targetElement = resultsArrayMemberObject.get("target");
+        final JsonObject targetObject = targetElement.getAsJsonObject();
+        logger.debug(String.format("targetType: %s", targetObject.get("type").toString()));
+        final JsonElement targetTypeElement = targetObject.get("type");
+        final String targetTypeValue = targetTypeElement.getAsString();
+        if (!"RULE".equals(targetTypeValue)) {
+            logger.debug(String.format("This is not a rule; skipping it. (It's a %s)", targetTypeValue));
+            return Optional.empty();
+        }
+        final JsonElement ruleElement = targetObject.get("rule");
+        final JsonObject ruleObject = ruleElement.getAsJsonObject();
+        logger.debug(String.format("ruleClass: %s", ruleObject.get("ruleClass").toString()));
+        final JsonElement ruleClassElement = ruleObject.get("ruleClass");
+        final String ruleClassValue = ruleClassElement.getAsString();
+        if (!"haskell_cabal_library".equals(ruleClassValue)) {
+            logger.debug(String.format("This is not a haskell_cabal_library rule; skipping it. (It's a %s rule)", ruleClassValue));
+            return Optional.empty();
+        }
+        return Optional.of(ruleObject);
+    }
+
+    private void addDependencyToGraph(final MutableDependencyGraph dependencyGraph, final JsonObject ruleObject) throws IntegrationException {
+        final JsonElement attributeElement = ruleObject.get("attribute");
+        final JsonArray attributeArray = attributeElement.getAsJsonArray();
+
+        final NameVersion dependencyNameVersion = extractDependencyDetails(attributeArray);
+        final Dependency artifactDependency = haskageCompNameVersionToDependency(dependencyNameVersion.getName(), dependencyNameVersion.getVersion());
+        try {
+            logger.debug(String.format("Adding %s to graph", artifactDependency.getExternalId().toString()));
+            dependencyGraph.addChildToRoot(artifactDependency);
+        } catch (final Exception e) {
+            logger.error(String.format("Unable to create dependency from %s/%s", dependencyNameVersion.getName(), dependencyNameVersion.getVersion()));
+        }
+    }
+
+    private NameVersion extractDependencyDetails(final JsonArray attributeArray) throws IntegrationException {
```

Comment:
I'd like to suggest more separation from parsing from transforming - you could have this Step use a Parser where the parser takes the string and returns say List<HaskellDependency> and then the Step knows how to transform them into a dependency that is added to the graph, you are close as it is. I'd think extractHaskellCabalLibraryRule does all the json parsing and returns a well formed HaskellDependency and instead of returning a JsonObject that later work is done to, do all the json work up front and return a pojo. Then addDependencyToGraph doesn't take a JsonObject but takes in the HaskellDependency pojo. Then in your tests you can have a test that checks JsonObject -> HaskellDependency works and a test that List<HaskellDependency> -> Graph. Would let you have smaller more isolated tests.

Reply:
done.